AR Connection Management Guide

jjb opened the pull request:
A new guide for AR Connection management, coming out of discussion in #27759. see rendered markdown here: https://github.com/jjb/rails/blob/ar-connection-management-guide/guides/source/active_record_connection_management.md

This is my first pass with everything relevant I could think of. I've dealt with these issues in various ways over the years and I might be forgetting details or use cases. Please give me any and all feedback and I'll try to improve the document!

one question i have right off: am i remembering this correctly?
> "In previous version of rails, you needed to think about managing connections when spawning threads, but this is no longer the case."

jjb committed:
AR Connection Management Guide

gaurish commented on guides/source/active_record_connection_management.md:
```diff
@@ -0,0 +1,107 @@
+===================================
+
+In general, Active Record will automagically manage database connections in an
+effecient manner. See [Configuring a Database](/configuring.html#configuring-a-database) and [Database pooling](configuring.html#database-pooling) for information about the basics of optimally configuring your application to use available resources.
```
efficient

gaurish commented on guides/source/active_record_connection_management.md:
```diff
@@ -0,0 +1,107 @@
+* Configuring a multi-process web server
+* Spawning processes
+
+In previous version of rails, you needed to think about managing connections when
```
for clarity, it would be useful to mention exact rails version since when this is no longer the case.

jjb replied:
I don't know when this was and some research wasn't able to find it. Maybe someone else knows?

gaurish reviewed (commented):
Looks good!

However, should this we a separate guide or a seperate in the existing guide?

eugeneius commented on guides/source/active_record_connection_management.md:
```diff
@@ -0,0 +1,107 @@
+### Before forking
+
+Before the server forks, you need to disconnect the pool so that the forked processes
+don't get confused. Here's how to do that in a puma config:
```
"get confused" is too vague - we should describe the problem(s) that are avoided by disconnecting the pool.

robin850 commented:
Hi @jjb,

Thanks for the pull request ! Such guide would be effectively interesting. This pull request has been put under the 5.2 milestone as there most likely going to be improvements regarding connection management and particularly regarding multiple connection support.

I take the liberty to cross-link to [this gist](https://gist.github.com/robin850/ce867eedbfccb327d637e274c9a622cc) that might not be totally true but that can be relevant in the future once multiple connection support is present.

arthurnn commented on guides/source/active_record_connection_management.md:
```diff
@@ -0,0 +1,107 @@
+* Spawning processes
+
+In previous version of rails, you needed to think about managing connections when
+spawning threads, but this is no longer the case.
```
I think this is a bit confusing.
You still need to change the pool size depending on the number of threads you have.

jjb replied:
@arthurnn when i said "when spawning threads", i mean when creating new threads in your application code. regarding that, my statement is correct.

but it seems you thought i meant "you need to consider the number of threads your server is running when you decide how many connections to put in the pool".

if you have a suggestion how to clarify that it is the former and not the latter, let me know!

(or if i'm not understanding your feedback, let me know that too :-D)

arthurnn commented on guides/source/active_record_connection_management.md:
````diff
@@ -0,0 +1,107 @@
+my_worker.rb. In this script, you are going to start two long-running processes.  Here's how you would go about doing that:
+
+```ruby
+ActiveRecord::Base.connection_pool.disconnect!
````
Maybe you should mention that you won't be able to use the connection on the parent process anymore after this.
As you are disconnecting it.

arthurnn reviewed (commented):
Left two small comments.

betesh commented on guides/source/active_record_connection_management.md:
```diff
@@ -0,0 +1,107 @@
+
+When spawning processes, you must do a small amount of manual database pool management.
+
+Let's say you you have a script that you are going to run as a working using `rails runner`,
```
`as a worker` ?

jjb committed:
make most text < 90 chars

whitespace only

jjb committed:
responding to PR feedback

clarifications
typo fixes

jjb committed:
note about the need to disconnect parent pool

jjb commented:
@gaurish

> However, should this we a separate guide or a seperate in the existing guide?

in #27759, @arthurnn said that a standalone guide sounds like a good idea

jjb commented:
@betesh @arthurnn @eugeneius @gaurish Thanks for the feedback! I addressed it all, take a look.

I also added a new section, [Discussion on relationship between parent and child pools.](https://github.com/jjb/rails/blob/ar-connection-management-guide/guides/source/active_record_connection_management.md#discussion-on-relationship-between-parent-and-child-pools)

hsgubert commented:
Hey @jjb, interesting guide. 

It took me a long time to understand how AR works, and a few months ago I wrote a general explanation about how connection managing plays out. It is a little more in depth than what you have written so far, but since people following this post are interested in the subject it might help:

[https://github.com/hsgubert/rails-sharding/wiki/How-rails-sharding-connection-handling-works](https://github.com/hsgubert/rails-sharding/wiki/How-rails-sharding-connection-handling-works)

@hmistry this could probably help in a few of your questions

hmistry commented:
@hsgubert Thank you for the link - it was very informative and helped me understand AR's connection architecture.

jjb commented:
Hi folks - thanks for your feedback and ideas!

I think I've stretched my expertise on this topic and won't be able to improve the guide much more. If someone from rails has specific change requests, let me know. Or if someone with more knowledge on this than I wants to fork my repo and do another PR, feel free!

jjb committed:
fix markdown list syntax

kddeisz commented on guides/source/active_record_connection_management.md:
```diff
@@ -0,0 +1,133 @@
+and [Database pooling](configuring.html#database-pooling) for information about the
+basics of optimally configuring your application to use available resources.
+
+Sometimes, you will need to manage Active Record connections and/or connecton
```
connection**

jjb replied:
fixed, thanks!

jjb committed:
fix typo

bquorning commented on guides/source/active_record_connection_management.md:
```diff
@@ -0,0 +1,133 @@
+
+Each process in a multi-process web server needs its own database connection pool.
+If the server is also multi-threaded, the threads within each process will automatically
+share that processes' connection pool.
```
I believe `processes'` is plural, and should be changed to either `process's` or `process'`.

bquorning commented on guides/source/active_record_connection_management.md:
````diff
@@ -0,0 +1,133 @@
+
+```ruby
+before_fork do
+  ActiveRecord::Base.connection_pool.disconnect!
````
Or `ActiveRecord::Base.clear_all_connections!` ?

matthewd commented:
Good news!... and bad news. :confused:

Between #31221, #31173, and #31241, I believe I've eliminated the need for all of this guidance: it should finally all Just Work without anyone needing to think about it.

jjb commented:
@matthewd all good news as far as i'm concerned :-D

Skimming those PRs I see words like "Improve" and "seem to be". Are we sure that 100% of what's in this guide is now irrelevant? If so then we can close this, and I'll maybe make a gist and tweet it around for folks working witih < 5.2(?) to find.

matthewd commented:
"Improve" is relative to the existing fork-safety measures, which already prevented connection re-use/stomping, and left us "only" with phantom closes.

And the seeming was more about whether #31221 is a reasonable solution for the full set of problems one might have wanted #23815 for.

Beyond that, my only hedging is in my feeling that some of these incantations have taken on mythical weight -- these changes address the last of the reasons *I know of* to add them.

---

A gist / blog post / etc somewhere sounds like a great idea! :heart:  (yep, < 5.2.)

jjb commented:
@matthewd Gotcha.

If you've looked through my guide and are sure none of it is relevant for the cases described (yay!), go ahead and close this.

jjb commented:
I've put my guide into this gist: https://gist.github.com/jjb/cacab45cbc5816238b633ed2228c8290
